quil-py/src/program/mod.rs

```diff
@@ -218,8 +225,75 @@ impl PyProgram {
             .to_owned())
     }
 
-    pub fn __str__(&self) -> String {
-        self.as_inner().to_string()
+    pub fn resolve_placeholders(&mut self) {
+        self.as_inner_mut().resolve_placeholders();
+    }
+
+    // TODO: Pre-resolve placeholders into a map so a non-panic error can be raised?
+    #[args("*", label_resolver = "None", qubit_resolver = "None")]
+    pub fn resolve_placeholders_with_custom_resolvers(
+        &mut self,
+        label_resolver: Option<Py<PyFunction>>,
+        qubit_resolver: Option<Py<PyFunction>>,
+    ) {
+        #[allow(clippy::type_complexity)]
+        let rs_qubit_resolver: Box<dyn Fn(&QubitPlaceholder) -> Option<u64>> =
+            if let Some(resolver) = qubit_resolver {
+                Box::new(move |placeholder: &QubitPlaceholder| -> Option<u64> {
+                    Python::with_gil(|py| {
+                        let resolved_qubit =
+                            resolver.call1(py, (placeholder.to_python(py).unwrap(),));
```

Comment:
Any concern about using `unwrap` here (and elsewhere)? or is this covered by your comment elsewhere about pyo3 wrapping this in a nice Python exception?

Reply:
Yup, this would also be covered by pyo3 wrapping mechanisms for panics. However, I will change this one to an `expect` because the expectation for it to not fail isn't obvious, but `to_python` here should never fail for a `QubitPlaceholder`